proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,282 @@
+
+
+### Use Site
+
+The Game module can access the helper API `run` since it is in the same package as Engine.
+
+```
+[Game]
+
+import Engine
+
+public func play() {
+    MainEngine().run() // Can access `run` as it is a package symbol in the same package
+}
+```
+
+However, if a client outside of the package tries to access the helper API, it will not be allowed.
+
+```
+[App]
+
+import Game
+import Engine
+
+let engine = MainEngine()
+engine.run() // Error: cannot find `run` in scope
+
+```
+### Package Names
+
+A package name is a unique string with c99 identifier characters, and is passed down to Swift frontend via a new flag `-package-name`.  It is then stored in the module binary and used to compare with package names of other modules to determine if they are part of the same package.  
+
+If `-package-name` is not given, the `package` access modifier is disallowed.  Swift code that does not use `package` access will continue to build without needing to pass in `-package-name`.
+
+Here's an example of how a package name is passed to a commandline invocation.
+
+```
+swiftc -module-name Engine -package-name gamePkg ...
+[Game] swiftc -module-name Game -package-name gamePkg ...
+[App] swiftc App -package-name appPkg ...
+```
+
+When building the Engine module, the package name 'gamePkg' and package symbols is stored in the Engine binary.  When building Game, the input to its package name 'gamePkg' is compared with Engine's package name; since they match, access to package symbols is allowed.  When building App, the name comparison shows 'appPkg' is different from its dependency module's so access to package symbols is deined, which is what we want.
+
+Swift Package Manager has a package identity per package, an identifier that's verified to be unique via a registry, and it will pass it down automatically.  Other build systems such as Bazel can introduce a new build setting to set the value for a package name.  Since it needs to be unique, a reverse-DNS name could be used to avoid clashing.
+
+### Exportability
+
+Package symbols will be stored in .swiftmodule only, and not in the public interface file (.swiftinterface).  We plan to introduce a .package.swiftinterface that contains package symbols, similar to a .private.swiftinterface which contains SPI symbols.  We will store information on whether the .swiftmodule was built from a .package.swiftinterface into the .swiftmodule.  We will also store the package name in both .swiftmodule and .swiftinterface.  This is to enforce loading a .swiftmodule instead of .swiftinterface when building modules in the same package.  
```

Comment:
This is very implementation-focused, and in theory Swift is an abstract language, not a specific implementation.  I'm not going to say this content is inappropriate for an evolution proposal, but you should try to be clear about the parts of the proposal that are specific to the primary Swift implementation.  I've made some suggestions below.

Also, I think this would be more appropriate in the previous section.

Also, please don't do this "we plan to" stuff.  If it's part of the proposal, it's part of the proposal.  If it's not, it's a future direction.

```suggestion
When the Swift frontend builds a `.swiftmodule` file directly from source, the file will include the package name and all of the `package` declarations in the module.  When the Swift frontend builds a `.swiftinterface` file from source, the file will include the package name, but it will put `package` declarations in a secondary `.package.swiftinterface` file.  When the Swift frontend builds a `.swiftmodule` file from a `.swiftinterface` file that includes a package name, but it does not have the corresponding `.package.swiftinterface` file, it will record this in the `.swiftmodule`, and it will prevent this file from being used to build other modules in the same package.
```